cumulus/client/cirrus-executor/src/lib.rs

```diff
@@ -395,12 +395,13 @@ where
 	/// Checks the execution receipt from the executor peers.
 	fn on_execution_receipt(
 		&self,
-		execution_receipt: &ExecutionReceipt<<Block as BlockT>::Hash>,
+		execution_receipt: &ExecutionReceiptFor<Block>,
 	) -> Result<Action, Self::Error> {
 		// TODO: validate the Proof-of-Election
 
 		let block_hash = execution_receipt.secondary_hash;
-		let block_number = self.client.expect_block_number_from_id(&BlockId::Hash(block_hash))?;
+		let block_number = execution_receipt.secondary_number;
```

Comment:
Why don't we extract block number from primary block hash though? I think we need to postpone block import or processing or ER until we know primary chain block on top of which ER was built, just like blockchain does when it sees block built on top of unknown parent.

Reply:
>I think we need to postpone block import or processing or ER until we know primary chain block on top of which ER was built

I agree. In Polkadot, the relay chain validators assigned to a parachain will try to achieve a BFT style agreement on the parachain block execution, if most of validators agree with the execution result, the parachain header will be included on the relay chain, only when the header had been included on the relay chain, the parachain will actually import that block. We lack such a thing right now, tracking the latest valid ER on the primary chain I mean. I think we need to figure out a scheme for this.

>Why don't we extract block number from primary block hash though?

Hmm, seems to be doable. I'll try and see.

Reply:
I _abuse_ the parachain consensus to retrieve the primary block number by primary hash within executor in https://github.com/subspace/subspace/pull/250/commits/814fbc0b6d5a0b71f5ad2cf5990ee7dda668718d, the code will be refactored eventually, but I'll skip it for now since we'll have a major upgrade later anyway.